File: js/checkers.py
import json
from js import console

# Game state
class CheckersGame:
    def __init__(self):
        self.reset()
    
    def reset(self):
        # Board representation:
        # 0 = empty
        # 1 = red piece, 2 = red king
        # -1 = black piece, -2 = black king
        self.board = [
            [0, 1, 0, 1, 0, 1, 0, 1],
            [1, 0, 1, 0, 1, 0, 1, 0],
            [0, 1, 0, 1, 0, 1, 0, 1],
            [0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0, 0],
            [-1, 0, -1, 0, -1, 0, -1, 0],
            [0, -1, 0, -1, 0, -1, 0, -1],
            [-1, 0, -1, 0, -1, 0, -1, 0]
        ]
        
        self.current_player = 1  # 1 = red, -1 = black
        self.selected_piece = None
        self.valid_moves = []
        self.must_jump = False
        self.game_over = False
        self.winner = None
        self.status = "Red's turn"
        console.log("Game reset, current_player =", self.current_player)
    
    def get_valid_moves(self, row, col):
        """Get all valid moves for a piece at given position"""
        piece = self.board[row][col]
        if piece == 0:
            return []
        
        # Check if piece belongs to current player
        if (piece > 0 and self.current_player == 1) or (piece < 0 and self.current_player == -1):
            moves = []
            jumps = []
            
            # Direction for regular pieces
            directions = []
            if piece == 1:  # Red piece moves down
                directions = [(1, -1), (1, 1)]
            elif piece == -1:  # Black piece moves up
                directions = [(-1, -1), (-1, 1)]
            elif piece in (2, -2):  # Kings move in all four diagonal directions
                directions = [(1, -1), (1, 1), (-1, -1), (-1, 1)]
            
            # Check each direction
            for dr, dc in directions:
                # Regular move
                new_row, new_col = row + dr, col + dc
                if 0 <= new_row < 8 and 0 <= new_col < 8 and self.board[new_row][new_col] == 0:
                    moves.append((new_row, new_col))
                
                # Jump move
                new_row, new_col = row + 2*dr, col + 2*dc
                mid_row, mid_col = row + dr, col + dc
                if (0 <= new_row < 8 and 0 <= new_col < 8 and 
                    self.board[new_row][new_col] == 0 and
                    ((self.board[mid_row][mid_col] > 0 and self.current_player == -1) or 
                     (self.board[mid_row][mid_col] < 0 and self.current_player == 1))):
                    jumps.append((new_row, new_col))
            
            # If jumps are available, only return jumps
            if jumps:
                return jumps
            else:
                return moves
        return []
    
    def check_for_jumps(self):
        """Check if there are any possible jumps for the current player"""
        for row in range(8):
            for col in range(8):
                piece = self.board[row][col]
                if (piece > 0 and self.current_player == 1) or (piece < 0 and self.current_player == -1):
                    moves = self.get_valid_moves(row, col)
                    # Check if any move is a jump (more than 1 square away)
                    for new_row, new_col in moves:
                        if abs(new_row - row) == 2:
                            return True
        return False
    
    def make_move(self, from_pos, to_pos):
        """Make a move from one position to another"""
        from_row, from_col = from_pos
        to_row, to_col = to_pos
        
        console.log(f"Making move from ({from_row},{from_col}) to ({to_row},{to_col})")
        
        # Move the piece
        piece = self.board[from_row][from_col]
        self.board[from_row][from_col] = 0
        self.board[to_row][to_col] = piece
        
        # Check for jump and remove captured piece
        if abs(to_row - from_row) == 2:
            mid_row = (from_row + to_row) // 2
            mid_col = (from_col + to_col) // 2
            console.log(f"Capturing piece at ({mid_row},{mid_col})")
            self.board[mid_row][mid_col] = 0
        
        # Check for king promotion
        if (piece == 1 and to_row == 7) or (piece == -1 and to_row == 0):
            console.log(f"Promoting piece at ({to_row},{to_col}) to king")
            self.board[to_row][to_col] = 2 if piece == 1 else -2
        
        # Check for additional jumps
        additional_jumps = False
        if abs(to_row - from_row) == 2:
            # Check if piece can jump again
            jumps = [move for move in self.get_valid_moves(to_row, to_col) if abs(move[0] - to_row) == 2]
            if jumps:
                console.log(f"Additional jumps available: {jumps}")
                additional_jumps = True
                self.selected_piece = (to_row, to_col)
                self.valid_moves = jumps
                self.must_jump = True
                return False  # Don't end turn yet
        
        if not additional_jumps:
            # End turn
            self.current_player *= -1
            self.selected_piece = None
            self.valid_moves = []
            console.log(f"Turn ended. New current player: {self.current_player}")
            
            # Check for game end conditions
            self.check_game_over()
            
            # Set status
            if self.game_over:
                if self.winner == 1:
                    self.status = "Game over! Red wins!"
                elif self.winner == -1:
                    self.status = "Game over! Black wins!"
                else:
                    self.status = "Game over! It's a draw!"
                console.log(f"Game over! Winner: {self.winner}")
            else:
                self.status = "Red's turn" if self.current_player == 1 else "Black's turn"
                
                # Check if current player must jump
                self.must_jump = self.check_for_jumps()
                if self.must_jump:
                    self.status += " (Must jump!)"
                    console.log("Current player must jump")
        
        return True  # Turn ended
    
    def check_game_over(self):
        """Check if the game is over (no valid moves or no pieces left)"""
        red_pieces = black_pieces = 0
        red_moves = black_moves = 0
        
        for row in range(8):
            for col in range(8):
                piece = self.board[row][col]
                if piece > 0:
                    red_pieces += 1
                    if self.current_player == 1:
                        red_moves += len(self.get_valid_moves(row, col))
                elif piece < 0:
                    black_pieces += 1
                    if self.current_player == -1:
                        black_moves += len(self.get_valid_moves(row, col))
        
        if red_pieces == 0:
            self.game_over = True
            self.winner = -1
        elif black_pieces == 0:
            self.game_over = True
            self.winner = 1
        elif (self.current_player == 1 and red_moves == 0) or (self.current_player == -1 and black_moves == 0):
            self.game_over = True
            self.winner = -1 if self.current_player == 1 else 1
            
    def handle_click(self, row, col):
        """Handle a click on the board at position (row, col)"""
        console.log(f"Python handle_click called: row={row}, col={col}")
        console.log(f"Current player: {self.current_player}, Selected piece: {self.selected_piece}")
        console.log(f"Valid moves: {self.valid_moves}")
        
        if self.game_over:
            console.log("Game is over, ignoring click")
            return False
            
        # If we already have a piece selected
        if self.selected_piece:
            selected_row, selected_col = self.selected_piece
            console.log(f"Have selected piece at ({selected_row}, {selected_col})")
            
            # If the clicked square is in the valid moves
            if (row, col) in self.valid_moves:
                console.log(f"Clicked on valid move square ({row}, {col})")
                # Make the move
                result = self.make_move((selected_row, selected_col), (row, col))
                console.log(f"Move made, result: {result}")
                return True
            # If clicking on a different piece of the same color and we're not in a must-jump situation
            elif self.board[row][col] * self.current_player > 0 and not self.must_jump:
                console.log(f"Clicked on different piece at ({row}, {col})")
                # Select the new piece
                self.selected_piece = (row, col)
                self.valid_moves = self.get_valid_moves(row, col)
                console.log(f"New valid moves: {self.valid_moves}")
                return True
            # If clicking on the same piece, deselect it if not in a must-jump situation
            elif row == selected_row and col == selected_col and not self.must_jump:
                console.log("Deselecting piece")
                self.selected_piece = None
                self.valid_moves = []
                return True
            else:
                console.log(f"Invalid click at ({row}, {col})")
        else:
            # If no piece is selected, try to select one
            piece = self.board[row][col]
            console.log(f"No piece selected. Clicked on square with piece value: {piece}")
            
            # Check if piece belongs to current player
            if (piece > 0 and self.current_player == 1) or (piece < 0 and self.current_player == -1):
                console.log("Piece belongs to current player")
                # Jumps are mandatory
                if self.must_jump:
                    console.log("Must jump is active")
                    moves = self.get_valid_moves(row, col)
                    jumps = [move for move in moves if abs(move[0] - row) == 2]
                    console.log(f"Available jumps: {jumps}")
                    if jumps:
                        self.selected_piece = (row, col)
                        self.valid_moves = jumps
                        return True
                    else:
                        console.log("No jumps available for this piece")
                else:
                    # Select the piece and calculate valid moves
                    self.selected_piece = (row, col)
                    self.valid_moves = self.get_valid_moves(row, col)
                    console.log(f"Selected piece at ({row}, {col}). Valid moves: {self.valid_moves}")
                    return True
            else:
                console.log("Piece does not belong to current player")
        
        console.log("No action taken")
        return False

# Initialize the game
game = CheckersGame()

# Functions exposed to JavaScript
def handle_click(row, col):
    # Validate row and column values
    if not (0 <= row < 8 and 0 <= col < 8):
        console.log(f"Invalid row/col values: row={row}, col={col}")
        return False
        
    result = game.handle_click(row, col)
    console.log(f"handle_click result: {result}")
    return result

def get_board_state():
    return json.dumps(game.board)

def get_game_status():
    status = game.status
    if game.selected_piece:
        row, col = game.selected_piece
        status += f" - Selected piece at ({row}, {col})"
    return status

def reset_game():
    game.reset()

console.log("Checkers game initialized") 

File: js/checkers.js
// Load the Pyodide interpreter and initialize the checkers game
let pyodideReadyPromise;

// Enable debug mode
const DEBUG_MODE = true;

// Log debug information
function debugLog(...args) {
  if (DEBUG_MODE) {
    console.log(...args);
  }
}

async function initializePyodide() {
  try {
    // Load Pyodide
    console.log("Starting to load Pyodide...");
    self.pyodide = await loadPyodide();
    console.log("Pyodide loaded successfully");
    
    await self.pyodide.loadPackagesFromImports(`
      import sys
      import js
    `);
    console.log("Python packages imported");
    
    // Load the checkers game from Python code
    console.log("Fetching Python code...");
    const pythonCode = await fetch('js/checkers.py')
      .then(response => {
        if (!response.ok) {
          throw new Error(`HTTP error! Status: ${response.status}`);
        }
        return response.text();
      });
    
    console.log("Python code fetched, running...");
    await self.pyodide.runPythonAsync(pythonCode);
    
    console.log("Pyodide initialized with checkers game");
    return self.pyodide;
  } catch (error) {
    console.error("Error initializing Pyodide:", error);
    document.getElementById('game-status').textContent = "Error loading game: " + error.message;
    throw error;
  }
}

// Start loading Pyodide right away
pyodideReadyPromise = initializePyodide();

// Function to check if Pyodide is loaded correctly
function isPyodideReady(pyodide) {
  if (!pyodide) {
    console.error("Pyodide is not initialized");
    return false;
  }
  
  try {
    // Test if we can access game object
    const testResult = pyodide.runPython('isinstance(game, CheckersGame)');
    debugLog("Pyodide ready check:", testResult);
    
    // Validate that the game object is properly initialized
    const validGame = validateGameState(pyodide);
    return testResult === true && validGame;
  } catch (error) {
    console.error("Error checking if Pyodide is ready:", error);
    return false;
  }
}

// Validate game state
function validateGameState(pyodide) {
  try {
    // Check if game has the expected attributes
    const hasBoard = pyodide.runPython('hasattr(game, "board")');
    const hasValidMoves = pyodide.runPython('hasattr(game, "valid_moves")');
    const hasSelectedPiece = pyodide.runPython('hasattr(game, "selected_piece")');
    
    debugLog("Game state validation:", { hasBoard, hasValidMoves, hasSelectedPiece });
    
    // Print the current board for debugging
    console.log("Current board state:");
    pyodide.runPython(`
      print("Board state:")
      for row in game.board:
          print(row)
      print("Current player:", game.current_player)
      print("Selected piece:", game.selected_piece)
      print("Valid moves:", game.valid_moves)
    `);
    
    return hasBoard && hasValidMoves && hasSelectedPiece;
  } catch (error) {
    console.error("Error validating game state:", error);
    return false;
  }
}

// Function to reload the Python module if needed
async function reloadPythonModuleIfNeeded() {
  try {
    debugLog("Checking if Python module needs to be reloaded...");
    
    // Try to access a method on the game object
    const methodsWorking = pyodide.runPython(`
      try:
          # Try calling a method to check if the object is still fully functional
          game.get_valid_moves(0, 0)
          True
      except Exception as e:
          print("Error calling method:", e)
          False
    `);
    
    if (!methodsWorking) {
      console.warn("Game object methods not working, reloading Python module...");
      
      // Reload the Python module
      const pythonCode = await fetch('js/checkers.py')
        .then(response => {
          if (!response.ok) {
            throw new Error(`HTTP error! Status: ${response.status}`);
          }
          return response.text();
        });
      
      await pyodide.runPythonAsync(pythonCode);
      console.log("Python module reloaded");
      return true;
    }
    
    debugLog("Python module working correctly");
    return false;
  } catch (error) {
    console.error("Error checking/reloading Python module:", error);
    return false;
  }
}

// Modify the syncStateWithPython function to reload if needed
async function syncStateWithPython() {
  try {
    debugLog("Syncing state with Python");
    
    // Check if module needs to be reloaded
    const reloaded = await reloadPythonModuleIfNeeded();
    if (reloaded) {
      console.log("Python module was reloaded, refreshing UI...");
    }
    
    // Get current state from Python
    const boardState = pyodide.runPython('get_board_state()');
    const hasSelectedPiece = pyodide.runPython('game.selected_piece is not None');
    const currentPlayer = pyodide.runPython('game.current_player');
    
    debugLog("Board state:", boardState);
    debugLog("Has selected piece:", hasSelectedPiece);
    debugLog("Current player:", currentPlayer);
    
    // Update UI
    updateBoard();
    updateStatus();
    updateVisualIndicators();
    
    debugLog("State sync complete");
  } catch (error) {
    console.error("Error syncing state:", error);
  }
}

// Initialize the game board and UI elements
window.addEventListener('DOMContentLoaded', async function() {
  const checkerboard = document.getElementById('checkerboard');
  const statusDisplay = document.getElementById('game-status');
  const resetButton = document.getElementById('reset-game');
  
  try {
    // Wait for Pyodide to be ready
    statusDisplay.textContent = "Loading Python interpreter...";
    const pyodide = await pyodideReadyPromise;
    statusDisplay.textContent = "Initializing game...";
    
    // Check if Pyodide is properly initialized
    if (!isPyodideReady(pyodide)) {
      statusDisplay.textContent = "Error: Failed to initialize game engine.";
      console.error("Pyodide ready check failed - game engine not properly initialized");
      return;
    }
    
    // Handle clicks on the board
    checkerboard.onclick = function(event) {
      // If clicked directly on a piece or indicator, find its parent square
      let target = event.target;
      console.log("Click target:", target.className);
      
      if (target.classList.contains('piece')) {
        console.log("Clicked on piece, finding parent square");
        target = target.parentElement;
      } else if (target.classList.contains('valid-move-indicator')) {
        console.log("Clicked on valid move indicator, finding parent square");
        target = target.parentElement;
      }
      
      // Only process if we clicked on a square
      if (target.classList.contains('square')) {
        const row = parseInt(target.dataset.row);
        const col = parseInt(target.dataset.col);
        console.log(`Processing click on square at row=${row}, col=${col}`);
        
        try {
          // Direct call to Python handle_click
          const result = pyodide.runPython(`handle_click(${row}, ${col})`);
          console.log("Python handle_click result:", result);
          
          // Update the board with current state
          updateBoard();
          updateStatus();
          
          // Show valid moves
          showValidMoves();
        } catch (error) {
          console.error("Error handling click:", error);
          statusDisplay.textContent = "Error: " + error.message;
        }
      }
    };
    
    // Set up the checkerboard
    function createBoard() {
      // Clear the board
      checkerboard.innerHTML = '';
      
      // Create the board squares
      for (let row = 0; row < 8; row++) {
        for (let col = 0; col < 8; col++) {
          const square = document.createElement('div');
          square.className = `square ${(row + col) % 2 === 0 ? 'light' : 'dark'}`;
          square.dataset.row = row;
          square.dataset.col = col;
          checkerboard.appendChild(square);
        }
      }
      
      // Update the board with pieces
      updateBoard();
    }
    
    // Update the visual board from the Python game state
    function updateBoard() {
      try {
        // Get the board state
        const boardState = pyodide.runPython('get_board_state()');
        console.log("Raw board state:", boardState);
        
        // Parse the board state
        let parsedState;
        try {
          parsedState = JSON.parse(boardState);
          console.log("Parsed board state:", parsedState);
        } catch (error) {
          console.error("Error parsing board state:", error);
          console.log("Invalid board state:", boardState);
          return;
        }
        
        // Remove all pieces, but keep squares and indicators
        document.querySelectorAll('.piece').forEach(el => el.remove());
        
        // Add pieces based on current state
        for (let row = 0; row < 8; row++) {
          for (let col = 0; col < 8; col++) {
            const value = parsedState[row][col];
            if (value !== 0) {
              const square = document.querySelector(`.square[data-row="${row}"][data-col="${col}"]`);
              if (!square) {
                console.error(`Could not find square at row=${row}, col=${col}`);
                continue;
              }
              
              const piece = document.createElement('div');
              
              if (value === 1) piece.className = 'piece red';
              else if (value === 2) piece.className = 'piece red king';
              else if (value === -1) piece.className = 'piece black';
              else if (value === -2) piece.className = 'piece black king';
              
              square.appendChild(piece);
            }
          }
        }
      } catch (error) {
        console.error("Error updating board:", error);
        statusDisplay.textContent = "Error updating board: " + error.message;
      }
    }
    
    // Update the game status display
    function updateStatus() {
      try {
        const status = pyodide.runPython('get_game_status()');
        console.log("Game status:", status);
        statusDisplay.textContent = status;
      } catch (error) {
        console.error("Error updating status:", error);
        statusDisplay.textContent = "Error updating status";
      }
    }
    
    // Reset the game
    resetButton.addEventListener('click', function() {
      try {
        console.log("Resetting game...");
        pyodide.runPython('reset_game()');
        updateBoard();
        updateStatus();
        
        // Clear any valid move indicators when resetting
        document.querySelectorAll('.valid-move-indicator').forEach(el => el.remove());
        document.querySelectorAll('.selected-piece').forEach(el => el.classList.remove('selected-piece'));
        
        console.log("Game reset successfully");
      } catch (error) {
        console.error("Error resetting game:", error);
        statusDisplay.textContent = "Error resetting game: " + error.message;
      }
    });
    
    // Update visual indicators for selected piece and valid moves
    function updateVisualIndicators() {
      try {
        // Remove any existing indicators
        document.querySelectorAll('.selected-piece').forEach(el => el.classList.remove('selected-piece'));
        document.querySelectorAll('.valid-move-indicator').forEach(el => el.remove());
        
        // Get selected piece
        const hasSelectedPiece = pyodide.runPython('game.selected_piece is not None');
        
        if (hasSelectedPiece) {
          // Highlight selected piece
          const selectedRow = pyodide.runPython('game.selected_piece[0]');
          const selectedCol = pyodide.runPython('game.selected_piece[1]');
          
          const selectedSquare = document.querySelector(`.square[data-row="${selectedRow}"][data-col="${selectedCol}"]`);
          if (selectedSquare && selectedSquare.firstChild) {
            selectedSquare.firstChild.classList.add('selected-piece');
          }
          
          // Highlight valid moves
          const validMovesStr = pyodide.runPython('json.dumps(game.valid_moves)');
          const validMoves = JSON.parse(validMovesStr);
          
          for (const [moveRow, moveCol] of validMoves) {
            const moveSquare = document.querySelector(`.square[data-row="${moveRow}"][data-col="${moveCol}"]`);
            if (moveSquare) {
              const indicator = document.createElement('div');
              indicator.className = 'valid-move-indicator';
              moveSquare.appendChild(indicator);
            }
          }
        }
      } catch (error) {
        console.error("Error updating visual indicators:", error);
      }
    }
    
    // Function to show valid moves for selected piece
    function showValidMoves() {
      try {
        // Remove any previous valid move indicators
        document.querySelectorAll('.valid-move-indicator').forEach(el => el.remove());
        
        // Remove previous selections
        document.querySelectorAll('.selected-piece').forEach(el => el.classList.remove('selected-piece'));
        
        // Check if there's a selected piece
        const hasSelectedPiece = pyodide.runPython('game.selected_piece is not None');
        console.log("Has selected piece:", hasSelectedPiece);
        
        if (hasSelectedPiece) {
          // Get the selected piece position
          const selectedRow = pyodide.runPython('game.selected_piece[0]');
          const selectedCol = pyodide.runPython('game.selected_piece[1]');
          
          // Highlight the selected piece
          const selectedSquare = document.querySelector(`.square[data-row="${selectedRow}"][data-col="${selectedCol}"]`);
          if (selectedSquare && selectedSquare.firstChild) {
            selectedSquare.firstChild.classList.add('selected-piece');
          }
          
          // Get and show valid moves
          const validMovesStr = pyodide.runPython('json.dumps(game.valid_moves)');
          console.log("Valid moves JSON:", validMovesStr);
          
          const validMoves = JSON.parse(validMovesStr);
          console.log("Valid moves:", validMoves);
          
          for (const move of validMoves) {
            const moveRow = move[0];
            const moveCol = move[1];
            console.log(`Adding indicator at ${moveRow},${moveCol}`);
            
            const moveSquare = document.querySelector(`.square[data-row="${moveRow}"][data-col="${moveCol}"]`);
            if (moveSquare) {
              const indicator = document.createElement('div');
              indicator.className = 'valid-move-indicator';
              moveSquare.appendChild(indicator);
            }
          }
        }
      } catch (error) {
        console.error("Error showing valid moves:", error);
      }
    }
    
    // Initialize the game
    createBoard();
    updateStatus();
    console.log("Game initialized successfully");
  } catch (error) {
    console.error("Error setting up game:", error);
    statusDisplay.textContent = "Error setting up game: " + error.message;
  }
});

// Expose a debug function that can be called from the console
window.debugCheckers = function() {
  try {
    console.log("Debugging checkers game state:");
    
    // Board state
    const boardState = pyodide.runPython(`
      print("Board state:")
      for row in game.board:
          print(row)
      game.board
    `);
    console.log("Board:", boardState);
    
    // Selected piece
    const selectedPiece = pyodide.runPython(`
      print("Selected piece:", game.selected_piece)
      game.selected_piece
    `);
    console.log("Selected piece:", selectedPiece);
    
    // Valid moves
    const validMoves = pyodide.runPython(`
      print("Valid moves:", game.valid_moves)
      game.valid_moves
    `);
    console.log("Valid moves:", validMoves);
    
    // Current player
    const currentPlayer = pyodide.runPython(`
      print("Current player:", game.current_player)
      game.current_player
    `);
    console.log("Current player:", currentPlayer);
    
    // Must jump flag
    const mustJump = pyodide.runPython(`
      print("Must jump:", game.must_jump)
      game.must_jump
    `);
    console.log("Must jump:", mustJump);
    
    return {
      boardState,
      selectedPiece,
      validMoves,
      currentPlayer,
      mustJump
    };
  } catch (error) {
    console.error("Error in debug function:", error);
    return null;
  }
}; 